# Tutorials

## RNA-seq

1. [fastq to counts](https://bigslu.github.io/tutorials/RNAseq/1.RNAseq_fastq.to.counts.html): Sequence quality filtering, alignment, and counts with [SEAsnake](https://bigslu.github.io/SEAsnake/vignette/SEAsnake_vignette.html): Software to our run pipeline
2. [Counts to voom](https://bigslu.github.io/tutorials/RNAseq/2.RNAseq_counts.to.voom.html): Count quality filtering and normalization with `edgeR` and `limma`
3. [Voom to DEG](https://bigslu.github.io/tutorials/RNAseq/3.RNAseq_voom.to.DEG.html): Linear modeling to determine differentially expressed genes (DEG) with `kimma`
4. [RNAseq modules](https://bigslu.github.io/tutorials/RNAseq/4.RNAseq_modules.html): Gene co-expression modules with `WGCNA`
5. [Gene set analysis](https://bigslu.github.io/tutorials/RNAseq/5.RNAseq_gene.set.analysis.html): Including hypergeometric enrichment, `enrichr`, gene set enrichment analysis (GSEA), and STRING networks

## AWS

1. [AWS account setup](https://bigslu.github.io/tutorials/AWS/1.AWS_setup_tutorial.html)
2. [Conda installation](https://bigslu.github.io/tutorials/AWS/2.AWS_conda_setup.html)
3. [R installation](https://bigslu.github.io/tutorials/AWS/2.AWS_R_setup.html)

## GitHub

1. [Contributing to R packages on GitHub](https://bigslu.github.io/tutorials/R_packages/contribute_github.html)
