---
title: "RNA-seq data analysis"
subtitle: "Co-expression modules (WGCNA)"
author: "Kim Dill-McFarland, kadm@uw.edu"
date: "version `r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
  pdf_document:
    toc: yes
    toc_depth: '3'
editor_options:
  chunk_output_type: console
urlcolor: blue
---

```{r include=FALSE}
knitr::opts_chunk$set(fig.height=2.5, fig.width = 8.5) 
```

# Overview

This document covers our recommended analysis pipeline to determine differentially expressed co-expression modules. This pipeline includes linear regression using `kimma` and module building using `WGCNA`. The example data are human, bulk, paired-end RNA-seq, but this pipeline can be applied to other organisms or single-read libraries.

# 0. Setup

## Software

This pipeline should be completed in [R](https://cran.r-project.org/) and [RStudio](https://www.rstudio.com/products/rstudio/download/). You should also install the following packages.

```{r eval=FALSE}
#CRAN packages
install.packages(c("tidyverse", "WGCNA"))

#Bioconductor packages
install.packages("BiocManager")
BiocManager::install(c("edgeR", "limma", "patchwork"))

#GitHub packages
install.packages("devtools")
devtools::install_github("BIGslu/BIGverse")
#Or if fails, install packages individually
devtools::install_github("BIGslu/kimma")
devtools::install_github("BIGslu/BIGpicture")
devtools::install_github("BIGslu/RNAetc")
devtools::install_github("BIGslu/SEARchways")
```

And load them into your current R session.

```{r warning=FALSE}
library(tidyverse)
library(BIGverse)
library(limma)
library(patchwork)
library(WGCNA)

set.seed(651)
```

## Example data

Example data were obtained from virus-stimulated human plasmacytoid dendritic cells. Actual patient identifiers and metadata have been altered for this tutorial.

> Dill-McFarland KA, Schwartz JT, Zhao H, Shao B, Fulkerson PC, Altman MC, Gill MA. 2022. Eosinophil-mediated suppression and Anti-IL-5 enhancement of plasmacytoid dendritic cell interferon responses in asthma. J Allergy Clin Immunol. Epub ahead of print. doi: [10.1016/j.jaci.2022.03.025](https://doi.org/10.1016/j.jaci.2022.03.025). --- [GitHub](https://github.com/altman-lab/P259_pDC_public)

Specifically, this tutorial uses RNAseq data processed using our [SEAsnake](https://bigslu.github.io/SEAsnake/vignette/SEAsnake_vignette.html) and [counts to voom](https://bigslu.github.io/tutorials/RNAseq/2.RNAseq_counts.to.voom.html) pipelines, resulting in voom-normalized, log2 counts per million (CPM) expression and associated sample metadata in a `limma EList` object in the `data_clean/` directory. These data are available in the `kimma` package within the `example.voom` data object.

# 1. Load data

All counts, gene, and sample metadata are contained in a single object in the `kimma` package. Note that these data contain only 1000 genes from the full data set create in our [counts to voom](pipeline2) pipeline. This is to reduce computational time.

```{r}
#Extract and rename data object
dat <- kimma::example.voom
```

# 2. Differential gene expression

We will take a supervised approach to co-expression module creation. Thus, we first determine genes that are associated with our variable(s) of interest. Then, these genes are grouped into modules using weighted gene correlation network analysis (WGCNA). You can also take an unsupervised approach using all genes in the data set. Please see the [WGCNA tutorial](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/) for an unsupervised example.

We take a supervised approach as it reduces computational time and results in more interpretable modules.

First, we determine genes associated with viral infection and/or asthma in this data set. You can see how we determined this as our best fit model in our [voom to DEG](https://bigslu.github.io/tutorials/RNAseq/3.RNAseq_voom.to.DEG.html) tutorial.

```{r}
virus_lme <- kmFit(dat = dat, 
                   model = "~ virus*asthma + (1|ptID)", 
                   run_lme = TRUE, 
                   use_weights = TRUE)
```

We extract genes significant for the interaction term as well as those significant for virus or asthma alone.

```{r}
genes.interact <- virus_lme$lme %>% 
  filter(variable == "virus:asthma" & FDR < 0.2) %>% 
  pull(gene)

genes.virus <- virus_lme$lme %>% 
  filter(variable == "virus" & FDR < 0.2) %>% 
  pull(gene)

genes.asthma <- virus_lme$lme %>% 
  filter(variable == "asthma" & FDR < 0.2) %>% 
  pull(gene)


genes.all <- unique(c(genes.interact, genes.virus, genes.asthma))
```

This results in `r length(genes.all)` genes for use in module building. Note that this example data set contains only 1000 genes out of the roughly 14,000 protein coding genes usually seen in an RNAseq data set. Thus, the genes used in module analysis here are many fewer than would be seen in a full data set.

We filter the `dat` object to contain only these genes.

```{r}
dat.subset <- dat[genes.all, ]
```

# 3. Module building

The goal of weighted gene correlation network analysis (WGCNA) is to reduce data complexity by grouping genes into co-expression modules. This can take 1000s of genes down to only 10s of modules for analysis, thus dramatically reducing multiple comparison penalties (e.g. FDR). WGCNA modules contain correlated genes with similar expression patterns across samples. However, genes in a module do not necessary have the same average expression and a module can, thus, contain both lowly and highly expressed genes.

## 3.1 Determine soft threshold

There are a number of parameters that impact module builds. The first we will explore is soft thresholding power, which determines how correlated genes must be in order to be in the same module. We check values from 1 to 30 in a signed network because positive/negative expression values have meaning.

```{r eval=FALSE}
# You may wish to run on multiple processors
# WGCNA::allowWGCNAThreads(nThreads = 4)

fit <- pickSoftThreshold(data = t(dat$E),
                         powerVector = c(1:30),
                         networkType = "signed")
```

```{r eval=FALSE, echo=FALSE}
#Multiple processor SOCK failed when knitting
save(fit, file = "results/sft.RData")
```

```{r echo=FALSE}
load("results/sft.RData")
fit$fitIndices
```

You can roughly see how gene correlation associates with power in the automatically generated table but we also create plots for easier interpretation. First looking at the signed R^2^, we see a general increase in gene correlation with power until correlation is maximized around power 17. In supervised module building, there are also often 1 or 2 local maxima such as powers 2 and 14 here.

```{r}
as.data.frame(fit$fitIndices) %>%
  #Calculate scale free topology
  dplyr::mutate(signed.R.sq = -sign(slope)*SFT.R.sq) %>% 
  ggplot(aes(x = Power, y = signed.R.sq, label = Power)) +
  geom_text(size = 3) +
  theme_classic() +
  labs(y = "Scale free topology model fit,signed R^2",
       x = "Soft threshold (power)") +
  scale_y_continuous(breaks = round(seq(0, 1, by = 0.1), digits = 1))
```

Next, we look at connectivity which is a measure of how many genes in a module meet the correlation cutoff. This roughly correlates with how many genes are in each module with lower connectivity resulting in smaller modules. Here, we see an expected decrease in connectivity with increase in power.

```{r}
as.data.frame(fit$fitIndices) %>%
  ggplot(aes(x = Power, y = mean.k., label = Power)) +
  geom_text(size = 3) +
  theme_classic() +
  labs(y = "Mean connectivity",
       x = "Soft threshold (power)")
```

Taken together, we choose a soft thresholding power of 14 because it is a local maximum and achieves correlation \> 0.8. We will also compare to power 23 as the true maximum correlation to show how this effects modules in this tutorial.

```{r}
as.data.frame(fit$fitIndices) %>% 
  filter(Power %in% c(14, 23))
```

### Why is my R^2^ so low?

If you've selected few genes and/or your data are not strongly correlated, you many not reach an R^2^ of 0.8 or your local maximum may be much lower. This is because there are not many genes that meet the correlation cutoff. In this case, we either choose the local maximum closest to 0.8 or go back and select more genes for module building.

## 3.2 Build modules

We build modules with our two chosen soft thresholding powers and WGCNA defaults, except we continue to use a signed network and ask it for numeric labels.

```{r}
mod.p14 <- blockwiseModules(datExpr = t(dat.subset$E),
                             power = 14,
                             networkType="signed",
                             numericLabels = TRUE,
                             nthreads = 4)

mod.p23 <- blockwiseModules(datExpr = t(dat.subset$E),
                             power = 23,
                             networkType="signed",
                             numericLabels = TRUE,
                             nthreads = 4)
```

The first thing to note is module 0. This is the trash module; genes in this module are *not* correlated with one another. In an unsupervised analysis with all genes, module 0 can't be analyzed as it is a random assortment of genes. In our supervised analysis, however, we already know that these genes are associated with virus and/or asthma, so we may be able to glean some insight from them (more on this in Section 4).

Comparing the two module builds, we see the power 14 has fewer genes in module 0 than power 23. This makes sense as power 23 demands that genes be more highly correlated in order to be in a module together. Since both builds have the same number of modules, the additional genes in module 0, power 23 come from those that "fell out" of all the other modules. In a power comparison, you may also see fewer modules at higher powers when enough genes "fall out" of the smaller modules to no longer meet the minimum module size, thus that entire module is redistributed across the other modules, mostly to module 0.

```{r}
as.data.frame(mod.p14$colors) %>% 
  rownames_to_column("gene") %>% 
  count(`mod.p14$colors`)

as.data.frame(mod.p23$colors) %>% 
  rownames_to_column("gene") %>% 
  count(`mod.p23$colors`)
```

## 3.3 Other module parameters

There are a lot of module building parameters as you can see in the `?blockwiseModules` help page. We like to explore `minModuleSize`, `deepSplit`, and `maxBlockSize` as these generally achieve useful modules. Here, we will use the power 14 build to assess these additional parameters.

### Minimum module size

As the name suggests, minimum module size is the minimum number of correlated genes needed to make a module. If you have a group of correlated genes that do not make the minimum, those genes get redistributed, thus changing the overall network.

The default `minModSize` is the minimum of 20 or 1/2 the number of genes. In these data, that is `r length(genes.all)/2` so our default run had `minModSize = 20` and we saw that our smallest module 8 was 35 genes.

If we change the minimum to 40 genes, we lose 3 modules, not just the former module 8 that was below this cutoff. This is because the removal of one module results in the redistribution of these genes, thus changing the overall network and shifting more than just the one small module.

```{r}
mod.p14.m40 <- blockwiseModules(datExpr = t(dat.subset$E),
                                power = 14,
                                networkType="signed",
                                numericLabels = TRUE,
                                nthreads = 4,
                                minModuleSize = 40)

as.data.frame(mod.p14.m40$colors) %>% 
  rownames_to_column("gene") %>% 
  count(`mod.p14.m40$colors`)
```

### Deep network split

The deep network split determines where the network can be split to form modules. It is a controlled parameter from 0 to 4 with being 0 least and 4 most sensitive. Thus, lower values result in less splitting and fewer modules. The default `deepSplit` is 2 so comparing to 0, we see fewer modules in our data.

```{r}
mod.p14.s0 <- blockwiseModules(datExpr = t(dat.subset$E),
                                power = 14,
                                networkType="signed",
                                numericLabels = TRUE,
                                nthreads = 4,
                                deepSplit = 0) 

as.data.frame(mod.p14.s0$colors) %>% 
  rownames_to_column("gene") %>% 
  count(`mod.p14.s0$colors`)
```

### Maximum block size

The maximum block size determines the largest group of genes analyzed. If you have more genes than the `maxBlockSize`, the data are pre-clusterd into blocks less than this size and modules are built from within these clusters. In general, lower `maxBlockSize` results in more modules and more genes in module 0 as a full network connecting all genes is not constructed and you may miss some connections.

Because we only have `r length(genes.all)` genes in this analysis, the default `maxBlockSize` of 5000 includes all genes. In most supervised module builds, this is appropriate and we rarely change this parameter. However, if you have highly variable gene expression, reducing the `maxBlockSize` can result in smaller, more interpretable modules.

For example, if we reduce the `maxBlockSize` to 200, we get more genes in module 0 and generally smaller modules. This is not recommended for these data and just shown as an example here.

```{r}
mod.p14.b200 <- blockwiseModules(datExpr = t(dat.subset$E),
                                power = 14,
                                networkType="signed",
                                numericLabels = TRUE,
                                nthreads = 4,
                                maxBlockSize = 200)

as.data.frame(mod.p14.b200$colors) %>% 
  rownames_to_column("gene") %>% 
  count(`mod.p14.b200$colors`)
```

### An Art, not a science

There are no absolute cutoffs in module building. This includes supervised gene selection and module build parameters.

Because we want to capture as much signal as possible, we use a relatively high FDR cutoff for gene selection in module building. You may go even higher than FDR \< 0.2 in your analysis. The exact cutoff is arbitrary and we generally aim for 1000 to 5000 genes for module building. This is, of course, not possible in the reduced example data set here. This range is rather large given the roughly 14,000 genes in the full data set, and it will likely take some trial and error in your data. In general, if you find that you have a lot of modules not associated with your variable(s) of interest, you may need to reduce the FDR, and thus the number of genes, selected for module building. Or you may leave those genes and simply ignore modules that are not significant.

Similarly, it is helpful to compare several module builds across different parameters as summarized below. You want to minimize genes in module 0 (because these are the uncorrelated genes) without creating too many or too small modules for analysis. This process has many "right" paths. Hence, the idea that it is an art, not a science.

```{r echo=FALSE}
data.frame(
  power =        c(14,  23,  14,  14,  14),
  minModSize =   c(20,  20,  40,  20,  20),
  deepSplit =    c(2,   2,   2,   0,   2),
  maxBlockSize = c(5000,5000,5000,5000,200),
  totalModules = c(max(mod.p14$colors),
                   max(mod.p23$colors),
                   max(mod.p14.m40$colors),
                   max(mod.p14.s0$colors),
                   max(mod.p14.b200$colors)),
  sizeModule0 =  c(sum(mod.p14$colors == 0),
                   sum(mod.p23$colors == 0),
                   sum(mod.p14.m40$colors == 0),
                   sum(mod.p14.s0$colors == 0),
                   sum(mod.p14.b200$colors == 0))
)
```

# 4. Differential module expression

Once you've finalized your modules, you can assess their expression similar to individual genes. Here, we use mean expression of genes in modules. You may use the module eigenvalues as well (contained in `mod.p14$MEs`). We generally see similar results with the two methods and chose mean module expression because the values are more directly interpretable in relation to actual expression.

## 4.1 Mean module expression

First, we calculate the mean expression of genes in each module.

```{r}
# List genes in modules and format module names
mod.genes <- as.data.frame(mod.p14$colors) %>%
  tibble::rownames_to_column("geneName") %>%
  dplyr::rename(module = "mod.p14$colors") %>%
  #add leading 0 to module names for correct sorting of factor
  dplyr::mutate(module = ifelse(module <= 9,
                                paste("0", module, sep=""),
                                module)) %>%
  #Add color var
  dplyr::mutate(mod.color = WGCNA::labels2colors(mod.p14$colors))

# Mean gene expression
mod.E <- as.data.frame(dat.subset$E) %>% 
  rownames_to_column("geneName") %>% 
  inner_join(mod.genes, by = "geneName") %>% 
  #Mean
  group_by(module) %>% 
  summarise(across(where(is.numeric), ~mean(., na.rm=TRUE)),
            .groups = "drop")
```

```{r}
mod.E
```

## 4.2 Linear modeling

We run a linear mixed effects model just like in Section 2, only now we give the mean gene expression for counts. We can use the original sample metadata as our library names are the same in the module data. Note that the gene-level weights cannot be applied to module data and thus, `use.weights = FALSE`.

```{r}
virus_mod <- kmFit(counts = mod.E,
                   meta = dat.subset$targets,
                   model = "~ virus*asthma + (1|ptID)", 
                   run_lme = TRUE, 
                   use_weights = FALSE)
```

We see that all modules are significant for virus and/or asthma, though none are signifcant for the interaction.

```{r}
summarise_kmFit(virus_mod$lme)
```

Thus, in a true analysis, we would remove the interaction in our final model.

```{r}
virus_mod2 <- kmFit(counts = mod.E,
                   meta = dat.subset$targets,
                   model = "~ virus + asthma + (1|ptID)", 
                   run_lme = TRUE, 
                   use_weights = FALSE)
```

```{r}
summarise_kmFit(virus_mod2$lme)
```

# 5. `RNAetc` custom module functions

Our package `RNAetc` contains functions to make module building easier. You can filter to genes of interest and test soft thresholds with `fit_modules`.

```{r}
fit <- fit_modules(dat = dat, 
                   genes = genes.all,
                   powerVector=c(1:30), 
                   networkType = "signed",
                   nThread = 4)
```

This outputs:

-   `dat`: Filtered `EList` containing only genes of interest
-   `genes`: Vector of genes of interest
-   `sft`: Soft thresholding results from `pickSoftThreshold`
-   `top.plot`: Scaled topology plot (shown below)
-   `connect.plot`: Connectivity plot (shown below)

```{r}
names(fit)
```

```{r}
fit$top.plot + fit$connect.plot
```

Then build modules with `make_modules` where you can give a specific soft thresholding value (`sft.value`) or ask it to determine that value based on the minimum acceptable R^2^ (`Rsq.min`). You can also set our commonly tested parameters `minModuleSize`, `deepSplit`, and `maxBlockSize` as well as determine is you want mean module expression (`mods.mean`), eigenvalues (`mods.eigen`), and/or DAVID formatted data frame of genes in modules (`david`).

```{r}
mod.p14 <- make_modules(fit,
                        sft.value = 14, # Or Rsq.min = 0.8
                        minModuleSize = 20, 
                        maxBlockSize = 500, 
                        deepSplit = 2,
                        networkType = "signed", 
                        mods.mean = TRUE, mods.eigen = TRUE, david = TRUE,
                        nThread = 4)
```

This results in:

-   `genes`, `sft`, `top.plot`, and `connect.plot` from `fit` plus
-   `mods`: data frame of genes in modules with cleaned names
-   `mods.mean`: Mean gene expression
-   `mods.eigen`: Eigenvalues of module expression
-   `david`: DAVID-formatted data frame of genes in modules

```{r}
names(mod.p14)

mod.p14$mods[1:3,]
mod.p14$mods.mean[1:3,1:3]
mod.p14$david[1:3, 1:3]
```

# 6. Annotating module function

See our [gene set analyses](https://bigslu.github.io/tutorials/RNAseq/5.RNAseq_gene.set.analysis.html) tutorial for methods to assign function to modules.

# R session

```{r}
sessionInfo()
```

# *Navigation*

-   Previous tutorial: [voom to DEG](https://bigslu.github.io/tutorials/RNAseq/3.RNAseq_voom.to.DEG.html)
-   Next tutorial: [gene set analyses](https://bigslu.github.io/tutorials/RNAseq/5.RNAseq_gene.set.analysis.html)
-   [Index](https://bigslu.github.io/tutorials/)

------------------------------------------------------------------------
